### ISSUE TITLE
[css-variables] CSS wide keywords in custom properties with dependencies

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#9131]
I was investigating a [Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1843783) about custom properties with css-wide keyword fallbacks.

The following test-case ([live](https://bug1843783.bmoattachments.org/attachment.cgi?id=9346498)) renders differently in Chromium and Safari. I can explain the difference, I believe.

Safari seems to first figure out what declarations apply, then you resolve dependencies between the values, and treat revert-layer as invalid at computed value time. This is similar to how longhands work.

Chromium seems to do something different, where their dependency-cycle detection also seems to deal with `revert-layer` somehow.

To me Safari's behavior seems easier to understand, but I might be biased because that's how Gecko implements custom-properties too. I _think_ Chromium's behavior is more correct tho, but a clarification on the spec and sanity-checking my understanding of the situation would be great.

@mdubet / @andruud is my understanding above right?

**Comment by emilio** [comment]
I think https://drafts.csswg.org/css-variables/#cycles means that what Gecko and WebKit are implementing seems a bit closer to what the spec says, I think?

The spec assumes one value per property, which is true if you do that... But then again the behavior difference between `--foo: var(--non-existing, unset)` and `--foo: unset` seems odd.

**Comment by Loirooriol** [comment]
I'm confused, you say Gecko behaves like WebKit, but I'm getting purple on both Gecko and Blink, while WebKit is blue.

**Comment by emilio** [comment]
Gecko behaves like WebKit as in "how we deal with custom properties". We didn't honor wide keywords from fallback on custom properties (there's a patch in the bug linked above to fix that).

**Comment by andruud** [comment]
> The following test-case ([live](https://bug1843783.bmoattachments.org/attachment.cgi?id=9346498)) [...]

You seem to be worried about cycle detection, but I don't see a cycle in that test? Maybe I'm just blind.

- On `.outer2`, `--foo` is reverted to the previous layer, and the computed value becomes `purple`.
- On `.outer2`, `--color` then also becomes `purple` via a successful substituion of `--foo`.

**Comment by emilio** [comment]
Yeah, I agree there's no cycle per se in that test, no, but `revert-layer` might introduce another edge in the graph, effectively.

The main difference in implementation is, I think, that Gecko/WebKit perform the cycle detection during substitution, _after_ we've computed the un-substituted value of all properties. Substitution can't (easily at least, given how we implement these) introduce new edges, because we've lost the information of the "potentially-reverted-to" declarations.

**Comment by Loirooriol** [comment]
I still think this is totally unrelated to cycles.

```html
<!doctype html>
<style>
  .outer {  --color: red }
  @layer {
    .inner {  --color: green  }
  }
  .inner {  --color: revert-layer;  color: var(--color)  }
</style>
<div class="outer">
  <div class="inner">
      Which color am I?
  </div>
</div>
```

Gecko & Blink: green
WebKit: red

**Comment by andruud** [comment]
@Loirooriol  Yeah, that may be.

Perhaps @emilio is concerned about a case like this, where a revert appears "late" and adds a dependency that's not otherwise there in the un-reverted value:

```html
<!doctype html>
<style>
  .outer { --x: blue; color: red;}
  @layer {
    .inner {
      --x: var(--y);
    }
  }
  .inner {
    /* --x and --y are cyclic via revert-layer */
    --x: var(--unknown, revert-layer);
    --y: var(--x, yellow);
    color: var(--y, green);
  }
</style>
<div class="outer">
  <div class="inner">
      Which color am I?
  </div>
</div>
```

**Comment by kizu** [comment]
I was playing a bit with the `revert-layer` in custom properties, and I did find that there is a possibility to introduce a cyclic dependency via layers as @andruud shows.

And there are for sure cases where we'd want the `--foo: var(--bar, revert-layer)` to work, I did attempt this a few times when doing experiments for [my latest article](https://kizu.dev/layered-toggles/) that utilizes cyclic dependencies.

I think I like how Chrome handles this the most currently: regardless of if a property is registered, I'd want the `revert-layer` in `--foo: var(--bar, revert-layer)` to revert the `--foo`.

When registering the property, it seems that in Safari we _can_ `revert-layer` a custom property, but only if it is registered with some specific syntax (not universal).

In Firefox Nightly, right now, we can't `revert-layer` regardless if we register property or not.

Here is a CodePen with an example: https://codepen.io/kizu/pen/MWRGoNp

(I think, _ideally_ it would be nice to have some way to pass around and apply the CSS-wide keywords through custom properties, but that's probably a bigger, separate issue)

**Comment by tabatkins** [comment]
Chrome's behavior, as far as I can tell, is what I'd naively expect to happen, particularly in [Ander's latest example](https://github.com/w3c/csswg-drafts/issues/9131#issuecomment-1660359090) where Chrome gives green. --x and --y are initially not cyclic, but fallback causes --x to change to a value that is cyclic with --y, so the final `var(--y, green)` triggers fallback too.
